src/tutorial/output.md

```diff
@@ -100,6 +100,31 @@ to make your (and your user's) life easier.
 
 </aside>
 
+
+<aside class="note">
+
+**Performance note:**
+Printing to the terminal is surprisingly slow!
+If you call things like `println!` in a loop,
+it can easily become a bottleneck in an otherwise fast program.
+
+To speed this up,
+it helps to acquire a lock on `stdout` (or `stderr`)
+and use `writeln!` to print to it directly:
+
+```rust
+use std::io::{self, Write};
+
+let stdout = io::stdout();
+let mut handle = stdout.lock();
+writeln!(handle, "foo: {}", 42)?;
+```
+
+Another trick would be to buffer the writing yourself.
+(By default the output gets flushed to the terminal on each newline.)
+
+</aside>
```

Comment:
This is a nice tip! But I think it's incomplete.

Firstly, if one wants to speed up stdout writing, acquiring the lock is probably not the biggest win. Using an `io::BufWriter` is probably the biggest win.

Secondly, the buffering story is a bit complex. For many typical command line programs, you probably want line buffering when printing to a terminal but standard block buffering via `io::BufWriter` when redirecting output to a file.

I don't know how best to express this succinctly though. If I had space for exactly one tip here though, I'd probably choose to show the use of `io::BufWriter` over acquiring a lock.

Reply:
Very interesting! I'd not have guessed that buffering is _much_ more of a win than locking.

Do you think this warrants being an actual section of the chapter instead of an aside?

> the buffering story is a bit complex

Yeah, that's why I only alluded to it in the last sentence… I could add easily add a line like `let mut handle = BufWriter::new(handle);` as an example, though.

Reply:
A lock that isn't under contention should be very fast. It probably won't even make any syscalls.

I'm not sure about adding a whole section on this. It's certainly something that has been on my mind lately, but I don't have a good feel for how important it is.

Reply:
In any case there should definitely be an example of how to use a Buffered output in there too (which would probably be more interesting than just the locking)